### DISCUSSION TITLE
Collaboration (Experimenting)

### DISCUSSION THREAD

**Discussion by fredchristianson** [fredchristianson/javascript-game-engine#1] [General]
I didn't think this through before suggesting it on Twitter.  We may be going through a lot of experimenting and changing as the project progresses.  Or maybe we'll find an expert who can straighten us out.

My goal is a flat community. Everyone can create branches and see each other's work.  There will be a few people who can merge to master (me for now).  See discussion on branches.

GitHub Discussions is probably the best place to communicate.  But we can try Twitter or Discord or anywhere else people want.    We don't need to all be part of all discussions.    Feel free to do what you want.  Let everyone know if we're all invited.

**Comment by chylinski82** [comment]
Hi All,

I'm here as much for the project as for the collaborative part, like 
using Github in a real team or perhaps other tools, like Trello. Not 
long ago i had my first interview for a coding job and when asked 
questions about team work, I didn't have any idea what are they talking 
about.

**Reply by Eddie7145** [reply]
Yeah I feel like you get to know more in the cause of this project

Basically it's just accommodating people, their ideas, opinions and learning to work hand in hand with them with goal of attaining something

**Comment by RimuhRimu** [comment]
Hi Fred and everyone, here's my thought. I'm not very experienced at contributing to projects, but this is a drawing that shows the workflow I've seen the most for big, medium, and small projects to be as organized as possible with the least effort in understanding GitHub. 

There are some other aspects that may be helpful later, like GitHub workflows to run a linter for everyone's code or something when someone pushes changes or opens a pull request that I've bumped into before.

This is a proposal since that way we avoid having to add everyone who wants to contribute to this projects by hand, they can do it freely just opening a pull request(providing a meaningful description of what they have done) from the changes from their fork, all we have left to be done is review what that person have done.

![image](https://user-images.githubusercontent.com/68438095/223931290-eb87f66b-3cda-465b-96d3-dbe843c07f59.png)

**Reply by fredchristianson** [reply]
Forks are a reasonable option and we can pull from them.  
I'm focused on branches so we can see how other people do things and give feedback.    
It's a little different than other projects since I'd encourage people to do the same thing for experience.  I'll try to build tasks that are like tutorial sites, but they will all build into a useful system.

Everyone's free to do things outside of the tasks in a fork or user branch (turn to feature branch for pull).

**Reply by chylinski82** [reply]
i guess i need to brush up on differences between forks and branches. I thought once you forked a repo, you could work on any branch, unless it was restricted.

**Reply by fredchristianson** [reply]
I need to brush up too.  I could be wrong.  

My understanding is that when you fork, you get your own repository.  Nobody can see the branches unless you do a pull request.

If create a branch on the original repository and push it, everyone can see it.    I'm encouraging this for anyone who wants feedback on their work.  Branches aren't just for implementing a feature that will make it to master.  Branches could get out of control, but we'll deal with that if it happens.  For now I think having "work-*" and "user-*" branches lets us have a lot of duplicate implementations without making feature and fix branches unusable.

Please correct me if I got something wrong.  I am a "beginner maintainer".
